### PR TITLE
Make GenLazyNativeFuncDefinition generator to be customizable in lazy codegen

### PR DESCRIPTION
**Pull request by wonjoo-wj**
As part of the ongoing LTC migration effort, PyTorch/XLA is updating its codegen to use `xla::Shape` instead of `torch::lazy::Shape`. To achieve this, this PR updates the codegen to make the `GenLazyNativeFuncDefinition` generator customizable. 

The existing `GenLazyNativeFuncDefinition` is kept by using the initial default values, so this change should not introduce any new behaviors to the existing codegen in PyTorch.

### CODE REVIEW

File: torchgen/gen_lazy_tensor.py
Comment (alanwaketan) [inline torchgen/gen_lazy_tensor.py:202]: It doesn't seem like we actually use this in XLA.

File: torchgen/gen_lazy_tensor.py
Comment (alanwaketan) [inline torchgen/gen_lazy_tensor.py:317]: `default_args. native_func_definition_generator`?

**Comment by alanwaketan** [comment]
Generally speaking, LGTM except the `shape_class` part.

**Comment by alanwaketan** [comment]
Also, you probably need to sign the CLA as an individual just to workaround it for now.

File: torchgen/gen_lazy_tensor.py
Comment (wonjoolee95) [reply torchgen/gen_lazy_tensor.py:317]: Updated!

File: torchgen/gen_lazy_tensor.py
Comment (wonjoolee95) [reply torchgen/gen_lazy_tensor.py:202]: Removed! This was a part of an older commit, but now it isn't needed since PyTorch/XLA overrides the entire `gen()` function. Good catch!

**Review by alanwaketan — APPROVED** [review]
LGTM. Thanks.

**Comment by wonjoolee95** [comment]
Thanks for the approval, Jiewen. Seems like we still need a `superuser` approval 😢 

@wconstab, any chance we can get a quick peek at this? Thanks!

**Comment by alanwaketan** [comment]
> Thanks for the approval, Jiewen. Seems like we still need a `superuser` approval 😢
> 
> @wconstab, any chance we can get a quick peek at this? Thanks!

Oh no, they rename codegen folder to torchgen... So our Lazy Tensor merge rules don't catch it. Let me do a pull request to update that.

**Review by wconstab — COMMENTED** [review]
I think I may as well stamp this to unblock you.  But I hesitate, since I know @bdhirsh and others have pointed out to me that this level of openness, providing whole classes of override/plugin to the codegen, makes it impossible for us to guarantee we won't break your stuff.  If possible, can xla::shape (name of class) be provided as a string input to the codegen instead?

**Comment by wonjoolee95** [comment]
Thanks for the comment, @wconstab. On XLA's side, the overriding for `NativeFunction` generator looks like https://github.com/pytorch/xla/blob/f571565ec226a5ce3a083a446864ac9739546ff5/scripts/gen_lazy_tensor.py#L159. We override the `shape_inference` and `build_ir_node` functions, so passing just a string `xla::Shape` won't be sufficient in this case. 

If we want to enforce the openness strictly, however, we can maybe add some kind of flag to the PyTorch codegen that can be used to have a specific logic/path for codegen for XLA only. This may not be very clean but can be a temporary solution while keeping the openness of the PyTorch codegen. 

cc @alanwaketan

**Review by wconstab — APPROVED** [review]
well, i'm gonna stamp for now. If we have to revisit it later we can look at all the customizations you've had to make and see how to factor them into a cleaner API.  And maybe it won't end up being a big issue.

**Comment by wonjoolee95** [comment]
> well, i'm gonna stamp for now. If we have to revisit it later we can look at all the customizations you've had to make and see how to factor them into a cleaner API. And maybe it won't end up being a big issue.

Sounds good. Thanks, Will. I'll keep a record of PyTorch lazy codegen customizations in the `[EXTERNAL] Lazy Tensor Core (LTC) Migration | Final Phases` design doc so we can revisit later if needed.

**Pull request closed without merging**
